### DISCUSSION TITLE
A way to write type-safe raw queries

### DISCUSSION THREAD

**Discussion by Newbie012** [prisma/prisma#4839] [Q&A]
I didn't invent anything, I just combined two epic libraries - Prisma and PgTyped. This is what I came up with:

## Write type-safe raw queries with Prisma and PgTyped

File structure:
 - src/
	 - **db/db.ts** *(responsible for initializing PrismaClient and a new pg client)*
	 - **context.ts** *(exposed both Prisma and a withConnection function - more on that later)*
	 - **db/queries/myQuery.ts** *(an example of a raw query - code below)*
	 - **db/queries/myQuery.types.ts** *(an auto generated file by pgtyped)*
	 
### db.ts
```ts
import { PrismaClient } from "@prisma/client";
import { Pool, PoolClient } from "pg";
import config from "../config";

// Initialize Prisma for basic querying operations
export const db = new PrismaClient();

// Initialize a pool for querying more complex queries
const pool = new Pool({ connectionString: config.DATABASE_URL });

pool.on("error", (err) => {
  console.error("Unexpected error on idle client", err);
  process.exit(-1);
});

// this function exposes the pool, while handling the connection itself
export async function withConnection<T>(callback: (poolClient: PoolClient) => Promise<T>) {
  let poolClient: PoolClient | null = null;

  try {
    poolClient = await pool.connect();
    const output = await callback(poolClient);
    return output;
  } catch (e) {
    console.error(e);
    throw e;
  } finally {
    if (poolClient) {
      poolClient.release();
    }
  }
}
```

### context.ts
```ts
import { PrismaClient } from "@prisma/client";
import { Client } from "pg";
import { withConnection } from "./db";

export interface Context {
  prisma: PrismaClient;
  withConnection<T>(callback: (client: Client) => Promise<T>): Promise<T>;
}

export const context: Context = {
  prisma: new PrismaClient(),
  withConnection: withConnection
};
```

### db/queries/myQuery.ts + myQuery.types.ts
This is where we write are raw query. At first, we will write something like:

```ts
import { sql } from  "@pgtyped/query";

export const getCampaignGeo = sql`
  SELECT *
  FROM campaign_geo
  LIMIT 1
`;
```

Currently, there's nothing special. The magic starts only after you run pgtyped:
(Make sure you have a configuration file. More on that on - https://pgtyped.now.sh/docs/getting-started)
```
npx pgtyped -w -c pgtyped.config.json
```

Once that's done, pgtyped has generated a types file (in my case my emitTemplate is `{{dir}}/{{name}}.types.ts`) called `db/queries/myQuery.types.ts`:
```ts
/** Types generated for queries found in "src/db/queries/myQuery.ts" */

/** 'GetCampaignGeo' parameters type */
export type IGetCampaignGeoParams = void;

/** 'GetCampaignGeo' return type */
export interface IGetCampaignGeoResult {
  id: number;
  campaignId: number;
  lat: number;
  lng: number;
  radius: number;
}

/** 'GetCampaignGeo' query type */
export interface IGetCampaignGeoQuery {
  params: IGetCampaignGeoParams;
  result: IGetCampaignGeoResult;
}
```
Once it's generated, I manually attach the interface `IGetCampaignGeoQuery` to `getCampaignGeo` this way:

```ts
import { sql } from  "@pgtyped/query";

export const getCampaignGeo = sql<IGetCampaignGeoQuery>`
  SELECT *
  FROM campaign_geo
  LIMIT 1
`;
```

### Now, for the fun part - mixing all together

Since I'm using nexus, I'll show my implementation that way:

```ts
export const queryRawField = queryField("queryRawField", {
  type: "...",
  args: {
    where: something, // { geoPoint: { lat: number, lng: number } }
  },
  async resolve(_root, args, ctx) {
    const campaignGeoId: number = await ctx.withConnection(async conn => {
      const result: IGetCampaignGeoResult[] = await getCampaignGeo.run(args.where, conn);
      return result[0]?.id;
    });

    return ctx.prisma.campaignGeo.findUnique({ where: { id: campaignGeoId }});
  },
});
```

And that's it.

**Comment by Newbie012** [comment]
I wrote a simple SQL query just to simplify things. In reality, of course, the queries that I'm writing are more complex

**Comment by trompx** [comment]
Hey @Newbie012, thanks a lot for sharing this.
I am looking to do something similar, the best would be to be able to directly execute raw queries inside prisma without losing the type safety.

Do you know if it is possible to use the prisma types directly in the template literal of tools like pgtyped/slonik/zapatos/postgres.js instead of having to generate them beforehand with the aforementioned tools?
And do you know if we can do what you did with pgtyped, but with postreges.js which just released the v3 version with great performance?
Are you still using this pattern?

**Reply by Newbie012** [reply]
Prisma and pgtyped doesn't have built in integration, but it doesn't really matter since they're both in typescript. I'm not using neither Prisma not pg-typed due to their lack of branded types. I've eventually created a library on my own that should work in conjunction with the library mfsqlchecker.

**Comment by alvis** [comment]
I found a community extension which maybe a perfect solution for this problem.
https://www.npmjs.com/package/prisma-extension-kysely

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
